src/Filters/remez_fir.jl

```diff
@@ -0,0 +1,829 @@
+                  linear phase filters.
+- `maxiter::Integer`: (optional)
+    Maximum number of iterations of the algorithm. Default is 25.
+- `grid_density:Integer`: (optional)
+    Grid density. The dense grid used in `remez` is of size
+    ``(numtaps + 1) * grid_density``. Default is 16.
+
+# Returns
+- `h::Array{Float64,1}`: A rank-1 array containing the coefficients of the optimal
+    (in a minimax sense) filter.
+
+[^McClellan1973a]: 
+J. H. McClellan and T. W. Parks, A unified approach to the
+design of optimum FIR linear phase digital filters,
+IEEE Trans. Circuit Theory, vol. CT-20, pp. 697-701, 1973.
+
+[^McClellan1973b]: 
+J. H. McClellan, T. W. Parks and L. R. Rabiner, A Computer
+Program for Designing Optimum FIR Linear Phase Digital
+Filters, IEEE Trans. Audio Electroacoust., vol. AU-21,
+pp. 506-525, 1973.
+
+# Examples
+Construct a length 35 filter with a passband at 0.15-0.4 Hz 
+(desired response of 1), and stop bands at 0-0.1 Hz and 0.45-0.5 Hz
+(desired response of 0). Note: the behavior in the frequency ranges between 
+those bands - the transition bands - is unspecified.
+
+```julia-repl
+julia> bpass = remez(35, [0 0.1 0.15 0.4 0.45 0.5], [0 1 0])
+```
+
+You can trade-off maximum error achieved for transition bandwidth. 
+The wider the transition bands, the lower the maximum error in the
+bands specified. Here is a bandpass filter with the same passband, but
+wider transition bands.
+
+```julia-repl
+julia> bpass2 = remez(35, [0 0.08 0.15 0.4 0.47 0.5], [0 1 0])
+```
+
+Here we compute the frequency responses and plot them in dB.
+
+```julia-repl
+using PyPlot
+b = DSP.Filters.PolynomialRatio(bpass, [1.0])
+b2 = DSP.Filters.PolynomialRatio(bpass2, [1.0])
+f = linspace(0, 0.5, 1000)
+plot(f, 20*log10.(abs.(freqz(b,f,1.0))))
+plot(f, 20*log10.(abs.(freqz(b2,f,1.0))))
+grid()
+```
+"""
+
+#=========
+Banner from C code
+
+-----------------------------------------------------------------------
+ SUBROUTINE: remez
+  THIS SUBROUTINE IMPLEMENTS THE REMEZ EXCHANGE ALGORITHM
+  FOR THE WEIGHTED CHEBYSHEV APPROXIMATION OF A CONTINUOUS
+  FUNCTION WITH A SUM OF COSINES.  INPUTS TO THE SUBROUTINE
+  ARE A DENSE GRID WHICH REPLACES THE FREQUENCY AXIS, THE
+  DESIRED FUNCTION ON THIS GRID, THE WEIGHT FUNCTION ON THE
+  GRID, THE NUMBER OF COSINES, AND AN INITIAL GUESS OF THE
+  EXTREMAL FREQUENCIES.  THE PROGRAM MINIMIZES THE CHEBYSHEV
+  ERROR BY DETERMINING THE BSMINEST LOCATION OF THE EXTREMAL
+  FREQUENCIES (POINTS OF MAXIMUM ERROR) AND THEN CALCULATES
+  THE COEFFICIENTS OF THE BEST APPROXIMATION.
+-----------------------------------------------------------------------
+=========#
+
+function remez(numtaps::Integer, bands::Vector, desired::Vector; 
+               weight::Vector=[], 
+               Hz::Real=1.0, 
+               filter_type::RemezFilterType=filter_type_bandpass,
+               maxiter::Integer=25, 
+               grid_density::Integer=16)
+    if length(weight)==0
+        weight = ones(desired)
```

Comment:
Could instead of this `if`-block  just use `fill(1.0, length(desired))` as the default value in function signature.